File: workdir/ctf/hackaday-u/session-4/exercises/script-exercises/ghidra-scripts/InstructionFrequency.py
# instruction_frequency.py
# @category MyScripts
# Count instruction frequency for function "main"

from ghidra.program.model.listing import Function

def find_function_by_name(name):
    fm = currentProgram.getFunctionManager()
    funcs = fm.getFunctions(True)  # iterate all functions
    for func in funcs:
        if func.getName() == name:
            return func
    return None

main_func = find_function_by_name("main")

if main_func is None:
    print("Error: Could not find 'main'")
else:
    listing = currentProgram.getListing()
    instr_iter = listing.getInstructions(main_func.getBody(), True)

    freq = {}

    while instr_iter.hasNext() and not monitor.isCancelled():
        instr = instr_iter.next()
        mnemonic = instr.getMnemonicString()
        freq[mnemonic] = freq.get(mnemonic, 0) + 1

    print("Instruction Frequency in 'main':")
    for mnem, count in freq.items():
        print("{} : {}".format(mnem, count))

File: workdir/ctf/hackaday-u/session-4/exercises/script-exercises/ghidra-scripts/InstructionFrequency.java
// InstructionFrequencySimple.java
// @category MyScripts
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.listing.Listing;

public class InstructionFrequency extends GhidraScript {

    @Override
    public void run() throws Exception {
        Function mainFunc = findFunction("main");
        if (mainFunc == null) {
            println("Cannot find 'main' function.");
            return;
        }

        Listing listing = currentProgram.getListing();
        InstructionIterator instrs = listing.getInstructions(mainFunc.getBody(), true);

        java.util.Map<String, Integer> freq = new java.util.HashMap<>();

        while (instrs.hasNext() && !monitor.isCancelled()) {
            Instruction instr = instrs.next();
            String mnem = instr.getMnemonicString();

            Integer count = freq.get(mnem);
            if (count == null) {
                freq.put(mnem, 1);
            } else {
                freq.put(mnem, count + 1);
            }
        }

        println("Instruction frequency in 'main':");
        for (String mnem : freq.keySet()) {
            println(mnem + ": " + freq.get(mnem));
        }
    }

    private Function findFunction(String name) {
        FunctionManager fm = currentProgram.getFunctionManager();
        FunctionIterator funcs = fm.getFunctions(true);
        while (funcs.hasNext()) {
            Function f = funcs.next();
            if (f.getName().equals(name)) {
                return f;
            }
        }
        return null;
    }
}
